### PR TITLE
Week 3: Add invariant and edge-case tests for LotteryAssignmentController

### PR DESCRIPTION
**Pull request by vipulpandey21**
## Description

`LotteryAssignmentController` (`esp/program/controllers/lottery.py`) is 1000+ lines with 20+ methods and no direct tests. The lottery runs once per program cycle and is irreversible — a silent regression in `compute_assignments()` or `fill_section()` cannot be corrected after it runs.

`LSRAssignmentTest` (already in `program/tests.py`) covers end-to-end enrolled/waitlist correctness and lunch constraints. This PR adds a complementary layer: direct invariant and edge-case tests that assert numpy array postconditions independently of the random seed, plus DB-level verification of `save_assignments()` and `clear_saved_assignments()`.

**4 test classes, 18 tests** in `esp/program/controllers/test_lottery_controller.py`:

`LotteryBasicTest` — initialization and basic operation:
- `compute_assignments()` completes without exception
- `student_sections` has correct shape and dtype
- `StudentSubjectInterest` records populate the interest matrix for valid sections
- `Priority/1` SRs populate the priority matrix
- Every student who expressed interest is either enrolled or has a verifiable blocking reason (section full, grade-blocked, or timeslot conflict)

`LotteryInvariantTest` — hard invariants on numpy arrays after `compute_assignments()`:
- No section has more enrolled students than its capacity
- No student is enrolled in two sections that share a timeslot
- No student enrolled outside their grade range when `check_grade=True`
- `student_schedules` is consistent with `student_sections`
- Students are only enrolled in sections they requested

`LotteryEdgeCaseTest` — edge cases:
- No SSI or Interested SR → interest matrix is all-False
- Only Priority/1 SRs → `compute_assignments()` completes correctly
- Fixed numpy seed → identical `student_sections` on two runs (regression guard for numpy upgrades)
- `clear_assignments()` resets arrays to zero

`LotterySaveTest` — DB behavior:
- `save_assignments()` creates `Enrolled` SRs matching `numpy.sum(student_sections)`
- Valid Enrolled SR count equals numpy sum
- `clear_saved_assignments()` expires all Enrolled SRs
- `clear_saved_assignments()` does not hard-delete records

## Related Issue

Refs #3780
Refs #794

## Type of Change
- [x] New feature

## Testing
- [x] Existing tests pass
- [x] New tests added (if applicable)
- [x] Manually verified

## AI Disclosure

Used Kiro (AI coding assistant) to verify this implementation.

## Checklist
- [x] Self-reviewed the code
- [ ] Updated documentation (if needed)
- [x] No new warnings or errors introduced

### CODE REVIEW

**Comment by vipulpandey21** [comment]
Hi Mr. Will and Mr. Ted, this covers both Weeks 3 and 4 from my plan  first-ever direct tests for LotteryAssignmentController. I know Mr. Ted prefers one problem per PR, but I've combined Weeks 3 and 4 here because both address the same single concern: LotteryAssignmentController having zero tests. Splitting them would have meant two PRs touching the same file with no logical boundary between them, so I felt one focused PR was cleaner.

Since the lottery is irreversible, I focused on the hard invariants that must hold regardless of random seed (no overcapacity, no timeslot conflicts, no grade violations), plus edge cases including a numpy upgrade regression guard, and DB-level verification of save_assignments() and clear_saved_assignments(). All 18 tests pass. Also addressed the two Copilot comments  removed an incorrect docstring claim and fixed the RNG state leak in the seed test. Would appreciate your review when you get a chance.

File: esp/esp/program/controllers/test_lottery_controller.py [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+            # Student got nothing — every requested section must have a blocking reason.
+            for sj_idx in numpy.nonzero(requested)[0]:
+                section_full = (
+                    numpy.sum(ctrl.student_sections[:, sj_idx])
+                    >= ctrl.section_capacities[sj_idx]
+                )
+                grade_blocked = ctrl.options['check_grade'] and (
+                    ctrl.student_grades[si_idx] < ctrl.section_grade_min[sj_idx]
+                    or ctrl.student_grades[si_idx] > ctrl.section_grade_max[sj_idx]
+                )
```
Comment (hwatheod) [inline esp/esp/program/controllers/test_lottery_controller.py:144]: Are these grade checks actually doing anything?  If I trace through the code correctly, it seems like [_make_class](https://github.com/learning-unlimited/ESP-Website/blob/7b7dc8f2aef51db32508b11e664fb3dcd51d1cf5/esp/esp/program/tests_catalog_sort.py#L19) assigns all classes to grades 7 to 12, while all the students you create here are in grade 9, so there can never be a conflict.  Similar question on `test_grade_range_respected` later in the code.

**Review by hwatheod — APPROVED** [review]
